### ISSUE TITLE
Colab GPU runtime ignores Graph execution error with `sparse_categorical_crossentropy`

### ISSUE THREAD

**Issue by rajdeep-biswas** [tensorflow/tensorflow#56979]
<details><summary>Click to expand!</summary> 
 
 ### Issue Type

Bug

### Source

source

### Tensorflow Version

2

### Custom Code

No

### Current Behaviour?

```shell
Setting output layer units to be a number that is lower than possible class labels, should make `sparse_categorical_crossentropy` throw a `Graph execution error` at the fit() step - which works as expected on normal runtime, however, switching the Google Colab runtime to GPU, the error is ignored and training proceeds as normal (with nan loss and 0 accuracy).
```

### Standalone code to reproduce the issue

```shell
GPU runtime should throw the same error as well instead of letting training proceed. I have created a well-documented notebook that can demonstrate / reproduce this issue and have [linked it here](https://colab.research.google.com/github/rajdeep-biswas/sparse-categorical-crossentropy/blob/master/sparcat.ipynb)..
```

### Relevant log output

_No response_</details>

**Comment by rajdeep-biswas** [comment]
[linking the notebook](https://colab.research.google.com/github/rajdeep-biswas/sparse-categorical-crossentropy/blob/master/sparcat.ipynb) that can reproduce the bug, here again because it might be easy to miss in the original PR comment.

**Comment by sushreebarsa** [comment]
@rajdeep-biswas I tried to replicate the issue, could you have a look at the gist here for both [gpu ](https://colab.research.google.com/gist/sushreebarsa/a18b8b90b7c024917bdc57dae69e5fe8/sparcat.ipynb#scrollTo=computational-issue)and [cpu](https://colab.research.google.com/gist/sushreebarsa/a96aeb1320b16ba6d03241f76ab27d59/sparcat.ipynb#scrollTo=computational-issue). Please confirm the gists. Thank you!

**Comment by rajdeep-biswas** [comment]
@sushreebarsa the gists look appropriate to me, thanks.

**Comment by gadagashwini** [comment]
I tried on Ubuntu with Tensorflow 2.9.1 and Cuda 11.4 (other than Colab GPU) still i could able to replicate the issue.
Tensorflow 2.9.1 and CUDA 11.2

```
Model: "sequential"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
=================================================================
 dense (Dense)               (None, 4)                 16        
                                                                 
 dense_1 (Dense)             (None, 18)                90        
                                                                 
=================================================================
Total params: 106
Trainable params: 106
Non-trainable params: 0
_________________________________________________________________
Epoch 1/10
1/1 [==============================] - 1s 990ms/step - loss: nan - accuracy: 0.0000e+00
Epoch 2/10
1/1 [==============================] - 0s 3ms/step - loss: nan - accuracy: 0.0000e+00
Epoch 3/10
1/1 [==============================] - 0s 3ms/step - loss: nan - accuracy: 0.0000e+00
Epoch 4/10
1/1 [==============================] - 0s 3ms/step - loss: nan - accuracy: 0.0000e+00
Epoch 5/10
1/1 [==============================] - 0s 2ms/step - loss: nan - accuracy: 0.0000e+00
Epoch 6/10
1/1 [==============================] - 0s 2ms/step - loss: nan - accuracy: 0.0000e+00
Epoch 7/10
1/1 [==============================] - 0s 2ms/step - loss: nan - accuracy: 0.0000e+00
Epoch 8/10
1/1 [==============================] - 0s 2ms/step - loss: nan - accuracy: 0.0000e+00
Epoch 9/10
1/1 [==============================] - 0s 2ms/step - loss: nan - accuracy: 0.0000e+00
Epoch 10/10
1/1 [==============================] - 0s 2ms/step - loss: nan - accuracy: 0.0000e+00
```

**Comment by reedwm** [comment]
This behavior is intentional, despite being inconsistent. This is documented in [`tf.nn.sparse_softmax_cross_entropy_with_logits`](https://www.tensorflow.org/api_docs/python/tf/nn/sparse_softmax_cross_entropy_with_logits):

> Each entry in `labels` must be an index in `[0, num_classes)`. Other values will raise an exception when this op is run on CPU, and return NaN for corresponding loss and gradient rows on GPU.

Unfortunately, it is very hard to raise errors for GPU ops based on the values of the input tensors. So in general, GPU ops do not raise errors even if the corresponding CPU ops do raise errors.

@fchollet, should this be documented in [`tf.keras.metrics.sparse_categorical_crossentropy `](https://www.tensorflow.org/api_docs/python/tf/keras/metrics/sparse_categorical_crossentropy) as well?
